stdlib/path.gr

```diff
@@ -0,0 +1,729 @@
+      let [first, ...rest] = tokens
+      if (fileType == File && !isFilePath(tokens)) {
+        None
+      } else {
+        let init = match (first) {
+          Slash => Some((Abs(Root), fileType, [])),
+          DriveTok(label) => Some((Abs(Drive(label)), fileType, [])),
+          _ => None,
+        }
+        Option.map(init => List.reduce(parseNextToken, init, rest), init)
+      }
+    },
+  }
+}
+
+// utility functions to translate path types
+
+let toTyped = (pathInfo: PathInfo) => {
+  let (base, fileType, subs) = pathInfo
+  (TBase(base), TFileType(fileType), subs): TypedPath<a, b>
+}
+
+let toUntyped = (typedPath: TypedPath<a, b>) => {
+  let (base, fileType, subs) = typedPath
+  let TBase(base) = base
+  let TFileType(fileType) = fileType
+  (base, fileType, subs): PathInfo
+}
+
+let pathInfo = (path: Path) => {
+  match (path) {
+    AbsoluteDir(x) => toUntyped(x),
+    AbsoluteFile(x) => toUntyped(x),
+    RelativeDir(x) => toUntyped(x),
+    RelativeFile(x) => toUntyped(x),
+  }
+}
+
+let toPath = (path: PathInfo) => {
+  match (path) {
+    (Abs(_), File, _) as p => AbsoluteFile(toTyped(p)),
+    (Abs(_), Dir, _) as p => AbsoluteDir(toTyped(p)),
+    (Rel(_), File, _) as p => RelativeFile(toTyped(p)),
+    (Rel(_), Dir, _) as p => RelativeDir(toTyped(p)),
+  }
+}
+
+// TODO expose these functions once module system added
+
+let absoluteFile = pathStr => {
+  let pathOpt = parseAbs(pathStr, File)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Absolute, File>>
+}
+
+let absoluteDir = pathStr => {
+  let pathOpt = parseAbs(pathStr, Dir)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Absolute, Directory>>
+}
+
+let parseRel = (pathStr, fileType) => {
+  let tokens = match (lexPath(pathStr)) {
+    [] => [Dot],
+    tokens => tokens,
+  }
+  let [first, ...rest] = tokens
+
+  if (fileType == File && !isFilePath(tokens)) {
+    None
+  } else {
+    let init = match (first) {
+      Dot => Some((Rel(0), fileType, [])),
+      Dotdot => Some((Rel(1), fileType, [])),
+      Text(str) => Some((Rel(0), fileType, [str])),
+      _ => None,
+    }
+    Option.map(init => List.reduce(parseNextToken, init, rest), init)
+  }
+}
+
+let relativeFile = pathStr => {
+  let pathOpt = parseRel(pathStr, File)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Relative, File>>
+}
+
+let relativeDir = pathStr => {
+  // parseRel(pathStr, Dir)//: Option<TypedPath<Relative, Directory>>
+  let pathOpt = parseRel(pathStr, Dir)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Relative, Directory>>
+}
+
+/**
+ * Parses a path string into a path wrapped with the path's type. If the path
+ * type is known in advance, the specialized parsing functions (`relativeDir`,
+ * `absoluteFile`, etc) should be used instead.
+ * 
+ * @param pathStr: The path string to parse
+ * @returns `Some(wrappedPath)` if the path is successfully parsed or `None` otherwise
+ * 
+ * @example fromString("/bin/") // an absolute Path referencing the directory /bin/
```

Comment:
Idea: What if these examples showed the type returned?
```suggestion
 * @example fromString("/bin/"): AbsoluteDir(TypedPath<Absolute, Directory>)
```
What do you think?

Reply:
Hmm, I was thinking of hiding `TypedPath` and keeping `Path` opaque until typed paths get added when these will become more relevant, although I could re-expose them if you think it'd be beneficial

Reply:
That's hard. I mostly want them exported so you can print a path and get the details. By abstracting them, you can't pattern match (but there isn't really anything you can do with the inner values)...

Reply:
Isn't it still possible to print the value even if it's not exported? I just tried it out and it looks like it still works.

But yeah, the lack of being able to do anything useful with them is why I figured it's not necessary to show right now.

Reply:
Unless I'm mistaken, printing an opaque type from outside the module will not print the representation.

Reply:
That's how it's supposed to work but we still have some bits to work out :P

Reply:
I just tried on my system and it looks like it does print the representation; is this a bug?

Reply:
Should I just keep it as is then?

Reply:
Sure